File: src/lib/api/tasks.ts
/**
 * Task APIs
 *
 * API functions for Task entity.
 */

import type { Task, CreateTaskParams, UpdateTaskParams } from "@/lib/types";
import { withTauriError } from "./utils";

export async function getTask(id: string): Promise<Task> {
  return withTauriError("get tasks", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<Task>("get_task", { id });
  });
}

export async function getTasks(): Promise<Task[]> {
  return withTauriError("get tasks", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<Task[]>("get_tasks");
  });
}

export async function getTasksByPlan(planId: string): Promise<Task[]> {
  return withTauriError("get tasks by plan", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<Task[]>("get_tasks_by_plan", { plan_id: planId });
  });
}

export async function createTask(data: CreateTaskParams): Promise<Task> {
  return withTauriError("create tasks", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<Task>("create_task", {
      plan_id: data.plan_id,
      title: data.title,
      description: data.description || null,
      start_date: data.start_date || null,
      end_date: data.end_date || null,
      priority: data.priority || null,
    });
  });
}

export async function updateTask(
  id: string,
  data: UpdateTaskParams,
): Promise<Task> {
  return withTauriError("update tasks", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<Task>("update_task", {
      id,
      title: data.title,
      description: data.description,
      start_date: data.start_date,
      end_date: data.end_date,
      status: data.status,
      priority: data.priority,
    });
  });
}

export async function deleteTask(id: string): Promise<void> {
  return withTauriError("delete tasks", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<void>("delete_task", { id });
  });
}

File: src-tauri/src/commands/tasks.rs
// Task CRUD commands

use crate::commands::repositories::TaskRepository;
use crate::log_command;
use crate::models::Task;
use crate::AppState;

#[tauri::command]
pub fn get_task(state: tauri::State<AppState>, id: String) -> Result<Task, String> {
    log_command!("get_task", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TaskRepository::get_by_id(&conn, &id)
    })
}

#[tauri::command]
pub fn get_tasks(state: tauri::State<AppState>) -> Result<Vec<Task>, String> {
    log_command!("get_tasks", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TaskRepository::get_all(&conn)
    })
}

#[tauri::command]
pub fn get_tasks_by_plan(
    state: tauri::State<AppState>,
    plan_id: String,
) -> Result<Vec<Task>, String> {
    log_command!("get_tasks_by_plan", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TaskRepository::get_by_plan(&conn, &plan_id)
    })
}

#[tauri::command]
pub fn create_task(
    state: tauri::State<AppState>,
    plan_id: String,
    title: String,
    description: Option<String>,
    start_date: Option<String>,
    end_date: Option<String>,
    priority: Option<String>,
) -> Result<Task, String> {
    log_command!("create_task", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        let id = uuid::Uuid::new_v4().to_string();
        TaskRepository::create(
            &conn,
            &id,
            &plan_id,
            &title,
            description.as_deref(),
            start_date.as_deref(),
            end_date.as_deref(),
            priority.as_deref(),
        )
    })
}

#[tauri::command]
pub fn update_task(
    state: tauri::State<AppState>,
    id: String,
    title: Option<String>,
    description: Option<String>,
    start_date: Option<String>,
    end_date: Option<String>,
    status: Option<String>,
    priority: Option<String>,
) -> Result<Task, String> {
    log_command!("update_task", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TaskRepository::update(
            &conn,
            &id,
            title.as_deref(),
            description.as_deref(),
            start_date.as_deref(),
            end_date.as_deref(),
            status.as_deref(),
            priority.as_deref(),
        )
    })
}

#[tauri::command]
pub fn delete_task(state: tauri::State<AppState>, id: String) -> Result<(), String> {
    log_command!("delete_task", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TaskRepository::delete(&conn, &id)
    })
}
